### PR TITLE
Dump routines for diag objects

### PR DESCRIPTION
**Pull request by rem1776**
**Description**
adds `dump_diag_obj` and `dump_diag_yaml_obj` routines that will dump the set fields of the objects for debugging. Uses helper routines that are private for the field and file obj output. Will also write to a log file if a name is given otherwise goes to stdout

**How Has This Been Tested?**
make check, intel 21

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: diag_manager/fms_diag_field_object.F90 [hunk 1]
```diff
@@ -964,5 +965,78 @@ PURE FUNCTION diag_field_id_from_name(this, module_name, field_name) &
...
+          write(unit_num, *) 'data_RANGE:' ,drange
+       type is(integer(i4_kind))
+          write(unit_num, *) 'data_RANGE:' ,drange
+       type is(integer(i8_kind))
+          write(unit_num, *) 'data_RANGE:' ,drange
+      end select
+    endif
+    write(unit_num, *) 'num_attributes:' ,this%num_attributes
+    if( allocated(this%attributes)) then
+      do i=1, SIZE(this%attributes)
```
Comment (uramirez8707) [inline diag_manager/fms_diag_field_object.F90:1021]: I think this should be `this%num_attributes`
this%attributes is allocated to be
https://github.com/NOAA-GFDL/FMS/blob/bd4c4c6ac6d7d0bcedcd92eca09a6033c15c9ee2/diag_manager/fms_diag_field_object.F90#L319

but the field can have less attributes than that

File: diag_manager/fms_diag_file_object.F90 [hunk 2]
```diff
@@ -587,6 +588,31 @@ subroutine add_start_time(this, start_time)
...
+
+  if( allocated(this%fileobj)) write( unit_num, *)'fileobj path', this%fileobj%path
+
+  write( unit_num, *)'type_of_domain', this%type_of_domain
+  if( allocated(this%file_metadata_from_model)) write( unit_num, *) 'file_metadata_from_model', &
+                                                                    this%file_metadata_from_model
+  if( allocated(this%field_ids)) write( unit_num, *)'field_ids', this%field_ids
+  if( allocated(this%field_registered)) write( unit_num, *)'field_registered', this%field_registered
+  if( allocated(this%num_registered_fields)) write( unit_num, *)'num_registered_fields', this%num_registered_fields
+  if( allocated(this%axis_ids)) write( unit_num, *)'axis_ids', this%axis_ids
```
Comment (uramirez8707) [inline diag_manager/fms_diag_file_object.F90:611]: you do `this%axis_ids(1:this%number_of_axis)`

**Review by uramirez8707 — COMMENTED** [review]
Thank you for doing this!

**Review by thomas-robinson — COMMENTED** [review]
THe interfaces for the dump routines are different.  Some take a filename and others are type bound and take a unit number.  I think these should all be the same?  Can the top routine call the get functions of the subobject?

File: diag_manager/fms_diag_file_object.F90 [hunk 3]
```diff
@@ -587,6 +588,31 @@ subroutine add_start_time(this, start_time)
 
 end subroutine
 
+!> writes out internal values for fmsDiagFile_type object
+subroutine dump_file_obj(this, unit_num)
```
[the remark this answers is not in the document]
Comment (rem1776) [reply diag_manager/fms_diag_file_object.F90:592]: Yes it's cause the `dump_file_obj` and `dump_field_obj` routines are just helper methods for `dump_diag_obj`.

It cycles through the field and file lists and writes them with their dump routines, while passing in the unit num so it's all written to the same place.

File: diag_manager/fms_diag_field_object.F90 (hunk 1 above)
Comment (rem1776) [reply diag_manager/fms_diag_field_object.F90:1021]: I'm pissed cause i had that originally lol, thought getting the allocated size would be better

**Review by uramirez8707 — APPROVED** [review]

**Pull request merged**